xdsl/transforms/jax_use_donated_arguments.py

```diff
@@ -0,0 +1,143 @@
+from dataclasses import dataclass
+
+from xdsl.context import MLContext
+from xdsl.dialects import builtin
+from xdsl.dialects.bufferization import MaterializeInDestinationOp
+from xdsl.dialects.builtin import ArrayAttr, DictionaryAttr, FunctionType, TensorType
+from xdsl.dialects.func import FuncOp, ReturnOp
+from xdsl.ir import BlockArgument, Operation, SSAValue
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+def map_donated_input_by_output(
+    donatable_inputs: Sequence[BlockArgument], outputs: Sequence[SSAValue]
+) -> dict[SSAValue, BlockArgument]:
+    """
+    Find suitable donated buffers for each of returned variables.
+    Each buffer can be used only once.
+    Types of the buffer and the variable should match.
+    """
+
+    donatable_inputs_by_type = {
+        type: tuple(group)
+        for type, group in itertools.groupby(
+            sorted(donatable_inputs, key=lambda i: str(i.type)), lambda i: i.type
+        )
+    }
+
+    nonbuffered_outputs = filter(
+        lambda out: not (
+            isinstance(out.owner, MaterializeInDestinationOp)
+            and isinstance(out.owner.dest, BlockArgument)
+        ),
+        outputs,
+    )
+    outputs_by_type = {
+        type: tuple(group)
+        for type, group in itertools.groupby(
+            sorted(nonbuffered_outputs, key=lambda i: str(i.type)), lambda i: i.type
+        )
+    }
+
+    mapping_by_type = {
+        k: tuple(zip(donatable_inputs_by_type[k], outputs_by_type[k]))
+        for k in donatable_inputs_by_type.keys() & outputs_by_type.keys()
+    }
+
+    return {o: i for mapping in mapping_by_type.values() for (i, o) in mapping}
+
+
+@dataclass
+class SubstituteDonatedTensors(RewritePattern):
+    """
+    Looks at returned tensors and if they match donated argument tensors ask bufferization to use them as buffers.
+    """
+
+    remove_matched_outputs: bool = False
+
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: ReturnOp, rewriter: PatternRewriter, /):
+        func_op = op.parent_op()
+        assert isinstance(func_op, FuncOp)
+
+        if func_op.arg_attrs is None or len(func_op.body.blocks) > 1:
+            return
+
+        donated_input_mask = tuple(
+            isinstance(inp.type, TensorType) and "tf.aliasing_output" in attr.data
+            for inp, attr in zip(func_op.args, func_op.arg_attrs, strict=True)
+        )
+        donated_inputs = tuple(itertools.compress(func_op.args, donated_input_mask))
+
+        if not donated_inputs:
+            return
+
+        donated_input_by_output = map_donated_input_by_output(
+            donated_inputs, op.arguments
+        )
+        if not donated_input_by_output:
+            return
+
+        ordered_buffered_outputs = tuple(
+            filter(lambda out: out in donated_input_by_output, op.arguments)
+        )
```

Comment:
why the filter? filter and map are slower and imo less readable than comprehensions
```suggestion
        ordered_buffered_outputs = tuple(
            arg for arg in donated_input_by_output if arg in donated_input_by_output
        )
```